File: 백준/Gold/14502. 연구소/연구소.py
n,m = map(int , input().split(" "))
graph = []
temp = [[0] * m for _ in range(n)]
dx = [-1,1,0,0]
dy = [0,0,-1,1]
for i in range(n):
    graph.append(list(map(int, input().split(" "))))
result = 0
#바이러스 퍼지게 하는 함수
def virus(x,y):
    for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]
        if 0<=nx<n and 0<=ny<m:
            if temp[nx][ny] == 0:
                temp[nx][ny] = 2
                virus(nx,ny)
#안전한 영역을 구하는 함수
def get_score():
    score = 0
    for i in range(n):
        for j in range(m):
            if temp[i][j] == 0:
                score +=1
    return score

def dfs(count):
    global result
    if count == 3:
        for i in range(n):
            for j in range(m):
                temp[i][j] = graph[i][j]

        for i in range(n):
            for j in range(m):
                if temp[i][j] == 2:
                    virus(i,j)
        #안전한 영역의 최댓값을 계속 넣어줌.
        result = max(result, get_score())
        return

    for i in range(n):

        for j in range(m):
  
            if graph[i][j]==0:
                graph[i][j] = 1
                # print(i,j)
                count+=1
                dfs(count)
                graph[i][j] = 0
                count-=1
dfs(0)
print(result)

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    private static final int wall = 3;
    private static int answer = 0;
    private static List<int[]> virus;
    private static int[] dx = new int[]{-1, 1, 0, 0};
    private static int[] dy = new int[]{0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int M = Integer.parseInt(input[1]);
        int[][] maps = new int[N][M];

        virus = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            String[] mapInput = br.readLine().split(" ");
            for (int j = 0; j < mapInput.length; j++) {
                maps[i][j] = Integer.parseInt(mapInput[j]);
                if(maps[i][j] == 2) {
                    virus.add(new int[]{i, j});
                }
            }
        }
        dfs(0, maps);
        System.out.println(answer);
    }

    static void dfs(int count, int[][] maps) {
        if (count == wall) {
            int[][] tempMap = new int[maps.length][maps[0].length];
            for (int i = 0; i < maps.length; i++) {
                for (int j = 0; j < maps[i].length; j++) {
                    tempMap[i][j] = maps[i][j];
                }
            }
            tempMap = checkVirus(tempMap);
            answer = Math.max(answer, getAnswer(tempMap));
            return;
        }

        for (int i = 0; i < maps.length; i++) {
            for (int j = 0; j < maps[i].length; j++) {
                if (maps[i][j] == 0) {
                    maps[i][j] = 1;
                    count+=1;
                    dfs(count, maps);
                    maps[i][j] = 0;
                    count -= 1;
                }
            }
        }
    }

    static int[][] checkVirus(int[][] maps) {
        boolean[][] check = new boolean[maps.length][maps[0].length];
        LinkedList<int[]> queue = new LinkedList<>();
        for (int i = 0; i < virus.size(); i++) {
            queue.add(virus.get(i));
            while (!queue.isEmpty()) {
                int[] poll = queue.poll();
                int x = poll[0];
                int y = poll[1];
                for (int j = 0; j < 4; j++) {
                    int nx = x + dx[j];
                    int ny = y + dy[j];

                    if (nx >= 0 && nx < maps.length && ny >= 0 && ny < maps[0].length) {
                        if (maps[nx][ny] == 0) {
                            maps[nx][ny] = 2;
                            queue.add(new int[]{nx, ny});
                        }
                    }
                }
            }
        }
        return maps;
    }

    static int getAnswer(int[][] maps) {
        int count = 0;
        for (int i = 0; i < maps.length; i++) {
            for (int j = 0; j < maps[i].length; j++) {
                if (maps[i][j] == 0) {
                    count++;
                }
            }
        }
        return count;
    }
}